File: src/validate/dashboard_is_valid.rs
use serde_json::Value;
use tokio::fs;

use crate::bundle_struct::Bundle;

pub async fn run(base: &str, bundle: &Bundle) -> Result<(), String> {
    let full_path = format!("{base}/{}", bundle.dashboard.path);

    let content = match fs::read_to_string(&full_path).await {
        Ok(v) => v,
        Err(e) => {
            return Err(format!(
                "ERROR: {}.{} Failed to read file_path={full_path}: error={e}",
                file!(),
                line!()
            ));
        }
    };

    let mut must_have: Vec<String> = vec![];
    must_have.push("__DASHBOARD_UUID__".to_string());
    must_have.push("__DATASOURCE__".to_string());
    must_have.push("__PROJECT_NAME__".to_string());

    for t in &bundle.tables {
        must_have.push(t.dashboard_var.clone());
    }

    for m in &must_have {
        if !content.contains(m) {
            return Err(format!(
                "ERROR: {}.{} Dashboard Must have {m} full_path={full_path}",
                file!(),
                line!()
            ));
        }
    }

    let dashboard_json: Value = match serde_json::from_str(&content) {
        Ok(v) => v,
        Err(e) => {
            return Err(format!(
                "ERROR: {}.{} Invalid JSON full_path={full_path} error={e}",
                file!(),
                line!()
            ));
        }
    };

    match dashboard_json["dashboard"].as_object() {
        Some(_) => (),
        None => {
            return Err(format!(
                "ERROR: {}.{} Invalid dashboard - top element must be dashboard.full_path={full_path}",
                file!(),
                line!()
            ));
        }
    }

    if dashboard_json["id"].as_str().is_some() {
        return Err(format!(
            "ERROR: {}.{} Invalid dashboard - cannot had Id set. full_path={full_path}",
            file!(),
            line!()
        ));
    }

    Ok(())
}

File: bundle-deployer/src/validation/dashboard_is_valid.ts
// Validation: Check that dashboard JSON is valid and contains required template variables

import type { Bundle } from "../types/bundle.ts";
import { getErrorMessage } from "../utils/error.ts";

export async function run(base: string, bundle: Bundle): Promise<void> {
  const dashboardPathList: string[] = [];
  
  dashboardPathList.push(`${base}/${bundle.dashboard.path}`);
  
  if (bundle.other_dashboards) {
    for (const otherDash of bundle.other_dashboards) {
      dashboardPathList.push(`${base}/${otherDash.path}`);
    }
  }
  
  for (const fullPath of dashboardPathList) {
    const content = await Deno.readTextFile(fullPath);
    
    // Build list of required template variables
    const mustHave: string[] = [
      "__DASHBOARD_UUID__",
      "__DATASOURCE__",
      "__PROJECT_NAME__",
    ];
    
    // Add all table dashboard_vars
    for (const table of bundle.tables) {
      mustHave.push(table.dashboard_var);
    }
    
    // Check that all required variables are present
    for (const variable of mustHave) {
      if (!content.includes(variable)) {
        throw new Error(
          `Dashboard must have ${variable} full_path=${fullPath}`
        );
      }
    }
    
    // Parse and validate JSON structure
    let dashboardJson;
    try {
      dashboardJson = JSON.parse(content);
    } catch (e) {
      throw new Error(
        `Invalid JSON full_path=${fullPath} error=${getErrorMessage(e)}`
      );
    }
    
    // Check that top element is "dashboard"
    if (!dashboardJson.dashboard || typeof dashboardJson.dashboard !== 'object') {
      throw new Error(
        `Invalid dashboard - top element must be dashboard. full_path=${fullPath}`
      );
    }
    
    // Check that id is not set
    if (dashboardJson.id !== undefined && dashboardJson.id !== null) {
      throw new Error(
        `Invalid dashboard - cannot have Id set. full_path=${fullPath}`
      );
    }
  }
}